ark/phenotyping/som_utils.py

```diff
@@ -0,0 +1,115 @@
+    """
+
+    # set fovs to all if None
+    if fovs is None:
+        fovs = img_xr.fovs.values
+
+    # set channels to all if None
+    if channels is None:
+        channels = img_xr.channels.values
+
+    # verify that the fovs and channels provided are valid
+    misc_utils.verify_in_list(fovs=fovs, image_fovs=img_xr.fovs.values)
+    misc_utils.verify_in_list(channels=channels, image_channels=img_xr.channels.values)
+
+    # define our flowsom matrix
+    flowsom_data = None
+
+    # iterate over fovs
+    for fov in fovs:
+        # subset img_xr with only the fov we're looking for, and cast to float32
+        img_data_blur = img_xr.loc[fov, ..., channels].values.astype(np.float32)
+
+        # for each marker, compute the Gaussian blur
+        for marker in range(len(channels)):
+            img_data_blur[:, :, marker] = ndimage.gaussian_filter(img_data_blur[:, :, marker],
+                                                                  sigma=blur_factor)
+
+        # flatten each image
+        pixel_mat = img_data_blur.reshape(-1, len(channels))
+
+        # convert into a dataframe
+        pixel_mat = pd.DataFrame(pixel_mat, columns=channels)
+
+        # assign metadata about each entry
+        pixel_mat['fov'] = fov
+        pixel_mat['row_index'] = np.repeat(range(img_data_blur.shape[0]), img_data_blur.shape[1])
+        pixel_mat['column_index'] = np.tile(range(img_data_blur.shape[0]), img_data_blur.shape[1])
+
+        # assign segmentation label
+        seg_labels_flat = seg_labels.loc[fov, ...].values.flatten()
+        pixel_mat['segmentation_label'] = seg_labels_flat
+
+        # remove any rows that sum to 0
+        pixel_mat = pixel_mat.loc[pixel_mat.loc[:, channels].sum(axis=1) != 0, :]
+
+        # normalize each row by total marker counts to convert into frequencies
+        pixel_mat.loc[:, channels] = pixel_mat.loc[:, channels].div(
+            pixel_mat.loc[:, channels].sum(axis=1), axis=0)
+
+        # assign to flowsom_data if not already assigned, otherwise concatenates
+        if flowsom_data is None:
+            flowsom_data = pixel_mat
+        else:
+            flowsom_data = pd.concat([flowsom_data, pixel_mat])
+
+    # normalize each marker column by the 99.9 percentile value
+    flowsom_data.loc[:, channels] = flowsom_data.loc[:, channels].div(
+        flowsom_data.loc[:, channels].quantile(q=0.999, axis=0), axis=1)
+
+    return flowsom_data
+
+
+def cluster_pixels(base_dir, chan_list):
+    """Run the FlowSOM training on the pixel data.
+
+    Saves results to pixel_mat_clustered.csv in base_dir.
+    Usage: Rscript som_runner.R {path_to_pixel_matrix} {chan_list_comma_separated} {save_path}
+
+    Args:
+        base_dir (str):
+            The path to the base directory
+        chan_list (list):
+            The list of markers to subset on
+    """
+
+    # get the path to som_runner.R
+    som_runner_path = os.path.dirname(os.path.realpath(__file__))
```

Comment:
What's the reasoning behind using os.path.realpath() here?

Reply:
1. This is because `som_runner.R` originally had to be run both by Travis and in Docker, and the directories are different. Now that we're not actually running `som_runner.R` in Travis, I think we can remove this and replace it with a hard coded path.
2. Can we can safely assume that the segmentation label tif file names will include the same prefixes before the underscore (`segmentationmask`) every time? It would make the `load_utils` calls a lot simpler in that case, and we can put both the channel and the segmentation TIFFs in the root of the Point directory.

Reply:
We can assume that the segmentation labels will be coming from our deepcell.org api call, and will have whatever format we want them to. We should mirror that setup for the example point here, or make any modifications to the directory structure we currently use for calling the deepcell server to make our life easier.